source/adios2/toolkit/burstbuffer/FileDrainer.cpp

```diff
@@ -0,0 +1,279 @@
+                 toOffset, countBytes);
+}
+void FileDrainer::AddOperationCopy(const std::string &fromFileName,
+                                   const std::string &toFileName,
+                                   size_t countBytes)
+{
+    AddOperation(DrainOperation::Copy, fromFileName, toFileName, 0, 0,
+                 countBytes);
+}
+
+void FileDrainer::AddOperationWriteAt(const std::string &toFileName,
+                                      size_t toOffset, size_t countBytes,
+                                      const void *data)
+{
+    std::string emptyStr;
+    AddOperation(DrainOperation::WriteAt, emptyStr, toFileName, 0, toOffset,
+                 countBytes, data);
+}
+
+void FileDrainer::AddOperationWrite(const std::string &toFileName,
+                                    size_t countBytes, const void *data)
+{
+    std::string emptyStr;
+    AddOperation(DrainOperation::Write, emptyStr, toFileName, 0, 0, countBytes,
+                 data);
+}
+
+void FileDrainer::AddOperationOpen(const std::string &toFileName, Mode mode)
+{
+    std::string emptyStr;
+    if (mode == Mode::Write)
+    {
+        AddOperation(DrainOperation::Create, emptyStr, toFileName, 0, 0, 0);
+    }
+    else if (mode == Mode::Append)
+    {
+        AddOperation(DrainOperation::Open, emptyStr, toFileName, 0, 0, 0);
+    }
+    else
+    {
+        throw std::runtime_error(
+            "ADIOS Coding ERROR: FileDrainer::AddOperationOpen() only supports "
+            "Write and Append modes\n");
+    }
+}
+
+InputFile FileDrainer::GetFileForRead(const std::string &path)
+{
+    auto it = m_InputFileMap.find(path);
+    if (it != m_InputFileMap.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        InputFile f = std::make_shared<std::ifstream>();
+        m_InputFileMap.emplace(path, f);
+        Open(f, path);
+        return f;
+    }
+}
+
+OutputFile FileDrainer::GetFileForWrite(const std::string &path, bool append)
+{
+    auto it = m_OutputFileMap.find(path);
+    if (it != m_OutputFileMap.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        OutputFile f = std::make_shared<std::ofstream>();
+        m_OutputFileMap.emplace(path, f);
+        Open(f, path, append);
+        return f;
+    }
+}
+
+void FileDrainer::Open(InputFile f, const std::string &path)
+{
+
+    f->rdbuf()->pubsetbuf(0, 0);
+    f->open(path, std::ios::in);
+}
+
+void FileDrainer::Open(OutputFile f, const std::string &path, bool append)
+{
+
+    if (append)
+    {
+        f->rdbuf()->pubsetbuf(0, 0);
+        f->open(path, std::ios::out | std::ios::app | std::ios::binary);
+    }
+    else
+    {
+        f->rdbuf()->pubsetbuf(0, 0);
+        f->open(path, std::ios::out | std::ios::trunc | std::ios::binary);
+    }
+}
+
+void FileDrainer::Close(InputFile f) { f->close(); }
+void FileDrainer::Close(OutputFile f) { f->close(); }
+
+bool FileDrainer::Good(InputFile f) { return (f->good()); }
```

Comment:
same here `InputFile& f`

Reply:
Why don't you suggest the same for the Seek() and Good() functions? They are in the protected area while Open/Close is in the private area. Does it make any difference?

Reply:
Sorry, missed those.  But yes, passing by reference (non-owning manner) should be used as much as possible.